Make null constraints consistent and use before_commit

parndt opened the pull request:
Fixes #20 by switching from `after_commit` to `before_commit` for
the operation that sets `created_by_id` and `updated_by_id` so
that we make sure to set it before committing it into postgresql. 

This also removes null constraints from any tables that use them
for `created_by_id` or `updated_by_id` with a migration.

parndt commented:
@tmfrnz per https://github.com/dumparkltd/marine-defrag-server/issues/20#issuecomment-968625221

we need to decide how to handle the seeds records because it currently doesn't let us run the seeds we have with the null constraints:

```
rails aborted!
ActiveRecord::NotNullViolation: PG::NotNullViolation: ERROR:  null value in column "created_by_id" of relation "taxonomies" violates not-null constraint
DETAIL:  Failing row contains (1, Convention status, null, 2021-11-15 20:54:20.802569, 2021-11-15 20:54:20.802569, t, null, f, null, null, null, null, null, null, null, null, null).
marine-defrag-server/db/seeds.rb:116:in `base_seeds!'
marine-defrag-server/db/seeds.rb:11:in `initialize'
marine-defrag-server/db/seeds.rb:525:in `new'
marine-defrag-server/db/seeds.rb:525:in `<top (required)>'

Caused by:
PG::NotNullViolation: ERROR:  null value in column "created_by_id" of relation "taxonomies" violates not-null constraint
DETAIL:  Failing row contains (1, Convention status, null, 2021-11-15 20:54:20.802569, 2021-11-15 20:54:20.802569, t, null, f, null, null, null, null, null, null, null, null, null).
marine-defrag-server/db/seeds.rb:116:in `base_seeds!'
marine-defrag-server/db/seeds.rb:11:in `initialize'
marine-defrag-server/db/seeds.rb:525:in `new'
marine-defrag-server/db/seeds.rb:525:in `<top (required)>'
Tasks: TOP => db:seed
```

tmfrnz commented:
not sure if i fully understand your question and/or context but

a) I don't think we generally need a `not null` constraint on 'updated_by_id` or `created_by_id` 
b) 'updated_by_id` or `created_by_id` fields are certainly not needed for all non-editable configuration records (https://github.com/dumparkltd/marine-defrag-server/issues?q=is%3Aissue+label%3AConfigRecord)

parndt commented:
@tmfrnz OK cool, I'll remove the columns `created_by_id` and `updated_by_id` from every table which has a label of `ConfigRecord` which is:

- `actortype_taxonomies`
- `actortypes`
- `measuretype_taxonomies`
- `measuretypes`
- `resourcestypes`
- `roles`
- `taxonomies`

Is that correct?

tmfrnz commented:
thanks @parndt that is pretty much correct at least the tables currently relevant for this project (except it should be `resourcetypes`). 

other non-editable config tables not currently relevant are
* frameworks
* framework_taxonomies

parndt commented:
Cool, thanks, I've updated the list because I'll refer to it later.

parndt commented on db/migrate/20211116065947_remove_null_constraints_from_created_by_id_and_updated_by_id.rb:
```diff
@@ -0,0 +1,15 @@
+  end
+
+  def down
+    change_column :actor_measures, :created_by_id, :bigint, null: true
+    change_column :measure_actors, :created_by_id, :bigint, null: true
```
```suggestion
    change_column :actor_measures, :created_by_id, :bigint, null: false
    change_column :measure_actors, :created_by_id, :bigint, null: false
```

parndt commented:
@tmfrnz I changed tack on this and removed the not null constraints because, you are right, they're not ultimately that helpful for us.

I also updated the seeds file to create the relationships directly instead of passing objects in. Let me know if you want me to revert this.

tmfrnz commented:
> @tmfrnz I changed tack on this and removed the not null constraints because, you are right, they're not ultimately that helpful for us.
> 
> I also updated the seeds file to create the relationships directly instead of passing objects in. Let me know if you want me to revert this.

thanks @parndt, looks great!

tmfrnz commented:
happy to confirm this worked on the dev environment
